File: xamarin/BrainBit/BrainBitDemo/BrainBitDemo/NeuroImpl/BrainBitController.cs
using System;
using System.Linq;
using System.Threading.Tasks;

using BrainBitDemo.Utils;

using NeuroSDK;

using Xamarin.Forms;

namespace BrainBitDemo.NeuroImpl;

internal class BrainBitController
{
#region Singleton
    private static BrainBitController _instance = null;
    public static  BrainBitController Instance { get => _instance ??= new BrainBitController(); }
#endregion

#region Search
    private Scanner        _scanner;
    private SensorsChanged _sensorFounded;

    public void StartSearch(SensorsChanged sensorChanged)
    {
        ISensorHelper sensorHelper = DependencyService.Get<ISensorHelper>();

        sensorHelper.EnableSensor(
            enabled =>
            {
                if (!enabled) return;

                _scanner ??= new Scanner(SensorFamily.SensorLEBrainBit, SensorFamily.SensorLEBrainBitBlack);

                _sensorFounded               =  sensorChanged;
                _scanner.EventSensorsChanged += _sensorFounded;
                _scanner.Start();
            }
        );
    }

    public void StopSearch()
    {
        _scanner?.Stop();
        if (_scanner != null) _scanner.EventSensorsChanged -= _sensorFounded;
        _sensorFounded = null;
    }
#endregion

#region Sensor
    private BrainBitSensor _sensor;

    public SensorStateChanged ConnectionStateChanged;
    public BatteryChanged     BatteryChanged;

    public SensorState ConnectionState { get => _sensor?.State ?? SensorState.StateOutOfRange; }

    public int BatteryPower { get => _sensor?.BattPower ?? 0; }

    public int SamplingFrequency { get => _sensor == null ? 0 : 250; }

    public Task<SensorState> CreateAndConnect(SensorInfo info)
    {
        SensorState outState = SensorState.StateOutOfRange;

        return Task<SensorState>.Factory.StartNew(
            () =>
            {
                try
                {
                    _sensor = _scanner.CreateSensor(info) as BrainBitSensor;

                    if (_sensor != null)
                    {
                        _sensor.EventSensorStateChanged += ConnectionStateChanged;
                        _sensor.EventBatteryChanged     += BatteryChanged;

                        outState = _sensor.State;
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e);
                }

                return outState;
            }
        );
    }

    public Task<SensorState> ConnectCurrent()
    {
        return Task<SensorState>.Factory.StartNew(
            () =>
            {
                if (_sensor.State != SensorState.StateOutOfRange) return SensorState.StateInRange;

                _sensor?.Connect();
                return _sensor.State;
            }
        );
    }

    public void DisconnectCurrent()
    {
        try
        {
            _sensor?.Disconnect();
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
        }
    }
#endregion

#region Signal
    public BrainBitSignalDataRecived SignalReceived;
    public BrainBitResistDataRecived ResistReceived;

    public void StartSignal()
    {
        _sensor.EventBrainBitSignalDataRecived += SignalReceived;
        _sensor?.ExecCommand(SensorCommand.CommandStartSignal);
    }

    public void StopSignal()
    {
        _sensor.EventBrainBitSignalDataRecived -= SignalReceived;
        _sensor?.ExecCommand(SensorCommand.CommandStopSignal);
    }

    public void StartResist()
    {
        _sensor.EventBrainBitResistDataRecived += ResistReceived;
        _sensor?.ExecCommand(SensorCommand.CommandStartResist);
    }

    public void StopResist()
    {
        _sensor.EventBrainBitResistDataRecived -= ResistReceived;
        _sensor?.ExecCommand(SensorCommand.CommandStopResist);
    }

#endregion

#region Info
    public string GetInfo()
    {
        if (_sensor?.State == SensorState.StateOutOfRange) return "Device unreachable!";

        var deviceInfo = "";

        var features = _sensor?.Features;
        deviceInfo += "\n===== [Features] =====\n";
        deviceInfo =  features?.Aggregate(deviceInfo, (current, feature) => current + $"[{feature}]: {_sensor.IsSupportedFeature(feature)}\n");

        var commands = _sensor?.Commands;
        deviceInfo += "\n===== [Commands] =====\n";
        deviceInfo =  commands?.Aggregate(deviceInfo, (current, command) => current + $" {command}\n");

        var parameters = _sensor?.Parameters;
        if (parameters == null) return deviceInfo;

        deviceInfo += "\n===== [Parameters] =====\n";
        foreach (ParameterInfo parameter in parameters)
        {
            switch (parameter.Param)
            {
            case SensorParameter.ParameterName:
                deviceInfo += $" Name: {_sensor.Name} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterState:
                deviceInfo += $" Connection state: {_sensor.State} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterAddress:
                deviceInfo += $" Address: {_sensor.Address} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterSerialNumber:
                deviceInfo += $" Serial number: {_sensor.SerialNumber} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterHardwareFilterState:
                break;
            case SensorParameter.ParameterFirmwareMode:
                deviceInfo += $" Firmware mode: {_sensor.FirmwareMode} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterSamplingFrequency:
                deviceInfo += $" Sampling frequency: {_sensor.SamplingFrequency} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterGain:
                deviceInfo += $" Gain: {_sensor.Gain} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterOffset:
                deviceInfo += $" Offset: {_sensor.DataOffset} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterFirmwareVersion:
                deviceInfo += $" Sensor version (Access: {parameter.ParamAccess}): "
                            + $"[FW]: {_sensor.Version.FwMajor}.{_sensor.Version.FwMinor} "
                            + $"[HW]: {_sensor.Version.HwMajor}.{_sensor.Version.HwMinor}.{_sensor.Version.HwPatch} "
                            + $"[Ext]: {_sensor.Version.ExtMajor}\n";
                break;
            case SensorParameter.ParameterBattPower:
                deviceInfo += $" (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterSensorFamily:
                deviceInfo += $"Sensor family: {_sensor.SensFamily} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterSensorMode:
                deviceInfo += $" Sensor mode (Access: {parameter.ParamAccess}): {_sensor.FirmwareMode}\n";
                break;
            case SensorParameter.ParameterExternalSwitchState:
            case SensorParameter.ParameterADCInputState:
            case SensorParameter.ParameterAccelerometerSens:
            case SensorParameter.ParameterGyroscopeSens:
            case SensorParameter.ParameterStimulatorAndMAState:
            case SensorParameter.ParameterStimulatorParamPack:
            case SensorParameter.ParameterMotionAssistantParamPack:
            case SensorParameter.ParameterMEMSCalibrationStatus:
            case SensorParameter.ParameterMotionCounterParamPack:
            case SensorParameter.ParameterMotionCounter:
            case SensorParameter.ParameterIrAmplitude:
            case SensorParameter.ParameterRedAmplitude:
            case SensorParameter.ParameterEnvelopeAvgWndSz:
            case SensorParameter.ParameterEnvelopeDecimation:
            case SensorParameter.ParameterSamplingFrequencyResist:
            case SensorParameter.ParameterSamplingFrequencyMEMS:
            case SensorParameter.ParameterSamplingFrequencyFPG:
            case SensorParameter.ParameterAmplifier:
            case SensorParameter.ParameterSensorChannels:
            case SensorParameter.ParameterSamplingFrequencyResp:
            case SensorParameter.ParameterSurveyId:
            case SensorParameter.ParameterFileSystemStatus:
            default:
                break;
            }
        }

        return deviceInfo;
    }
#endregion
}

File: react-native/brainbit2/NeuroImpl/BrainBitController.ts
import { Scanner, SensorInfo, SensorFamily, BrainBit2Sensor, SensorState, SignalChannelsData, ResistRefChannelsData, BrainBit2ChannelMode, EEGChannelType, SensorCommand, SensorFeature, SensorParameter, SensorParamAccess, SensorFirmwareMode, SensorDataOffset, SensorGain, SensorSamplingFrequency } from "react-native-neurosdk2";
import { PermissionsAndroid, Platform } from 'react-native';

let instance: BrainBitController;

class BrainBitController {

  private static _instance: BrainBitController;

  private constructor() {
    //...
  }

  public static get Instance() {
    // Do you need arguments? Make it a regular static method instead.
    return this._instance || (this._instance = new this());
  }

  private _scanner: Scanner | undefined

  private async requestPermissionAndroid() {
    try {
      const result = await PermissionsAndroid.requestMultiple([
        PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
        PermissionsAndroid.PERMISSIONS.BLUETOOTH_CONNECT,,
        PermissionsAndroid.PERMISSIONS.BLUETOOTH_SCAN,  // for android 12 (api 31+)
      ]);
      if (
        result[PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION] !=
        PermissionsAndroid.RESULTS.GRANTED
      ) {
        this.requestPermissionAndroid();
      }
      if (
        result[PermissionsAndroid.PERMISSIONS.BLUETOOTH_CONNECT] !=
        PermissionsAndroid.RESULTS.GRANTED
      ) {
        this.requestPermissionAndroid();
      }
    } catch (err) {
      console.warn(err);
    }
  }

  async startSearch(sensorFounded: (sensorFounded: SensorInfo[]) => void): Promise<void> {

    if (this._scanner != undefined) {
      this._scanner.close()
      this._scanner = undefined
    }
    this._scanner = new Scanner()
    await this._scanner.init([SensorFamily.LEBrainBit2, SensorFamily.LEBrainBitFlex, SensorFamily.LEBrainBitPro])
    this._scanner?.AddSensorListChanged(sensorFounded)
    if(Platform.OS != 'ios'){
      await this.requestPermissionAndroid()
    }
    await this._scanner?.start()
  }

  async stopSearch(): Promise<void> {
    this._scanner?.RemoveSensorListChanged()
    await this._scanner?.stop()
  }

  private _sensor: BrainBit2Sensor | undefined
  public connectionChangedCallback: ((state: SensorState)=>void) | undefined
  public batteryCallback: ((battery: number)=>void) | undefined

  public get connectionState(): SensorState {
    return this._sensor === undefined ? SensorState.OutOfRange : this._sensor.getState();
  }

  public get batteryPower(): number {
    return this._sensor === undefined ? 0 : this._sensor.getBattPower();
  }

  public get channelsCount(): number {
    return this._sensor?.getChannelsCount() ?? 0;
  }

  channelsList(): Array<String> {
    var result = Array(this._sensor?.getChannelsCount()).fill("")
    this._sensor?.getSupportedChannels()?.forEach((ch)=>{
      result[ch.Num] = ch.Name + ch.Num
    })

    return result
  }

  async createAndConnect(info: SensorInfo): Promise<SensorState> {
    return new Promise<SensorState>(async (resolve, reject) => {
      this._scanner?.createSensor(info)
        .then((sensor) => {
          this._sensor = sensor as BrainBit2Sensor
          this._sensor.AddConnectionChanged((state) => { 
            if(this.connectionChangedCallback != undefined)
              this.connectionChangedCallback(state); 
          })
          this._sensor.AddBatteryChanged((battery) => { 
            if(this.batteryCallback != undefined)
              this.batteryCallback(battery); 
          })
          if(this.connectionChangedCallback != undefined)
              this.connectionChangedCallback(SensorState.InRange);
          resolve(SensorState.InRange)
        })
        .catch((ex) => { reject(SensorState.OutOfRange) })
    });
  }

  async connectCurrent(): Promise<SensorState> {
    return new Promise<SensorState>(async (resolve, reject) => {
      if(this._sensor?.getState() != SensorState.OutOfRange) {
        resolve(SensorState.InRange)
        return
      }
      this._sensor?.connect()
        .then(() => {
          resolve(SensorState.InRange)
        })
        .catch((ex) => { reject(SensorState.OutOfRange) })
    });
  }

  async disconnectCurrent(): Promise<void> {
    await this._sensor?.disconnect()
  }

  public signalReceivedCallback: ((data: Array<SignalChannelsData>)=>void) | undefined
  public resistReceivedCallback: ((data: Array<ResistRefChannelsData>)=>void) | undefined

  async startSignal(){
    this._sensor?.AddSignalReceived((data)=>{
      if(this.signalReceivedCallback != undefined)
        this.signalReceivedCallback(data)
    })
    await this._sensor?.execute(SensorCommand.StartSignal).catch((ex)=> console.log(ex))
  }

  async stopSignal(){
    this._sensor?.RemoveSignalReceived()
    await this._sensor?.execute(SensorCommand.StopSignal).catch((ex)=> console.log(ex))
  }

  async startResist(){
    this._sensor?.AddResistanceReceived((data)=>{
      if(this.resistReceivedCallback != undefined)
        this.resistReceivedCallback(data)
    })
    await this._sensor?.execute(SensorCommand.StartResist).catch((ex)=> console.log(ex))
  }

  async stopResist(){
    this._sensor?.RemoveResistanceReceived();
    await this._sensor?.execute(SensorCommand.StopResist).catch((ex)=> console.log(ex))
  }

  get info(): string{
    if (this._sensor === undefined || this._sensor.getState() == SensorState.OutOfRange) return `Device unreachable!`;
        var deviceInfo = ``

        var features = this._sensor.getFeatures();
        deviceInfo += `Features:\n`
        features.forEach(feature => {
               deviceInfo += ` ${SensorFeature[feature]}\n`
        });

        var commands = this._sensor.getCommands()
        deviceInfo += `Commands:\n`
            commands.forEach(command => {
              deviceInfo += ` ${SensorCommand[command]}\n`
            });

        var parameters = this._sensor.getParameters()
        deviceInfo += 'Parameters:\n'
        parameters.forEach(parameter => { 
          switch(parameter.Param){
            case SensorParameter.Name:
              deviceInfo += ` Name (${SensorParamAccess[parameter.ParamAccess]}): ${this._sensor?.getName()}\n`
              break;
            case SensorParameter.State:
              deviceInfo += ` State (${SensorParamAccess[parameter.ParamAccess]}): ${SensorState[this._sensor === undefined ? 1 : this._sensor.getState()]}\n`
              break;
            case SensorParameter.Address:
              deviceInfo += ` Address (${SensorParamAccess[parameter.ParamAccess]}): ${this._sensor?.getAddress()}\n`
              break;
            case SensorParameter.SerialNumber:
              deviceInfo += ` Serial number: (${SensorParamAccess[parameter.ParamAccess]}): ${this._sensor?.getSerialNumber()}\n`
              break;
            case SensorParameter.FirmwareMode:
              deviceInfo += ` Firmware mode: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorFirmwareMode[this._sensor === undefined ? 1 : this._sensor.getFirmwareMode()]}\n`
              break;
            case SensorParameter.SamplingFrequency:
              deviceInfo += ` Sampling frequency: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorSamplingFrequency[this._sensor === undefined ? 10 : this._sensor.getSamplingFrequency()]}\n`
              break;
            case SensorParameter.Offset:
              deviceInfo += ` Offset: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorDataOffset[this._sensor === undefined ? 255 : this._sensor.getDataOffset()]}\n`
              break;
            case SensorParameter.FirmwareVersion:
              var version = this._sensor?.getVersion()
              deviceInfo += ` Firmware version: (${SensorParamAccess[parameter.ParamAccess]}):\n` +
                 `  FW: ${version?.FwMajor}.${version?.FwMinor}.${version?.FwPatch}` +
                 `  HW: ${version?.HwMajor}.${version?.HwMinor}.${version?.HwPatch} ` +
                 `  Ext: ${version?.ExtMajor}\n`
              break;
            case SensorParameter.BattPower:
              deviceInfo += ` Battery power: (${SensorParamAccess[parameter.ParamAccess]}): ${this._sensor?.getBattPower()}\n`
              break;
            case SensorParameter.SensorFamily:
              deviceInfo += ` Sensor family: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorFamily[this._sensor === undefined ? 0 : this._sensor.getSensFamily()]}\n`
              break;
          }
        });

        deviceInfo += '\n';
        deviceInfo += `  supported channels: [${this._sensor.getSupportedChannels().map((ch, i) => {
          return `\n   {Id: ${ch.Id}, Num: ${ch.Num}, Name: ${ch.Name}, Type: ${EEGChannelType[ch.ChType]}}`
        })}\n   ]\n`;

        var ampParams = this._sensor?.getAmplifierParam()

        deviceInfo += `  Amplifier params: \n` + 
        `   Current = ${ampParams.Current}\n` +
        `   ChResistUse = [ ${ampParams.ChResistUse.map((resist, i) => {
          return `\n      ${i}: ${resist} `
        })}\n   ]\n` +
        `   ChSignalMode = [ ${ampParams.ChSignalMode.map((mode, i) => {
          return `\n      ${i}: ${BrainBit2ChannelMode[mode]} `
        })}\n   ]\n` +
        `   ChannelGain = [ ${ ampParams.ChGain.map((gain, i) => {
          return `\n      ${i}: ${SensorGain[gain]} `
        })}\n   ]\n`;

        deviceInfo += '\n';

        return deviceInfo
  }
}

const BBControllerInstance = BrainBitController.Instance;
export default BBControllerInstance;